### ISSUE TITLE
Mismatch Between Training Progress and History/CSVLogger Callback Values

### ISSUE THREAD

**Issue by arthurflor23** [keras-team/keras#20212]
In Keras 3, the training metrics (accuracy and loss) recorded by the History callback don't match the progress printed during training. This seems to be an existing problem, as noted in the following related issues:

* [keras-team/keras#10014](https://github.com/keras-team/keras/issues/10014)
* [tensorflow/tensorflow#43470](https://github.com/tensorflow/tensorflow/issues/43470)
* [tensorflow/tensorflow#48408](https://github.com/tensorflow/tensorflow/issues/48408)

The issue appears when the training accuracy and loss logged during training differ from the final values stored in the History object or logged by CSVLogger. The validation metrics, however, are accurate!

Sample Code:
```
import keras
import numpy as np
import pandas as pd

X_train = np.random.random((100, 10))
y_train = np.random.randint(2, size=(100, 1))

X_val = np.random.random((20, 10))
y_val = np.random.randint(2, size=(20, 1))

model = keras.Sequential([
    keras.layers.Input(shape=(10,)),
    keras.layers.Dense(32, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

history = model.fit(X_train, y_train, epochs=10, batch_size=10, verbose=1, validation_data=(X_val, y_val), shuffle=False)

pd.options.display.float_format = '{:.4f}'.format

print("\nTraining History:")
print(pd.DataFrame(history.history))
```

Output:
```
Epoch 1/10
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.4851 - loss: 0.7064 - val_accuracy: 0.5500 - val_loss: 0.7063
Epoch 2/10
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.4965 - loss: 0.6964 - val_accuracy: 0.4000 - val_loss: 0.7035
Epoch 3/10
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5262 - loss: 0.6918 - val_accuracy: 0.5000 - val_loss: 0.7027
Epoch 4/10
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5601 - loss: 0.6892 - val_accuracy: 0.5500 - val_loss: 0.7029
Epoch 5/10
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5578 - loss: 0.6875 - val_accuracy: 0.5500 - val_loss: 0.7030
Epoch 6/10
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5698 - loss: 0.6863 - val_accuracy: 0.5500 - val_loss: 0.7033
Epoch 7/10
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5723 - loss: 0.6851 - val_accuracy: 0.5500 - val_loss: 0.7033
Epoch 8/10
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5723 - loss: 0.6839 - val_accuracy: 0.5500 - val_loss: 0.7035
Epoch 9/10
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5723 - loss: 0.6828 - val_accuracy: 0.5500 - val_loss: 0.7041
Epoch 10/10
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5930 - loss: 0.6816 - val_accuracy: 0.5500 - val_loss: 0.7046

Training History:
   accuracy   loss  val_accuracy  val_loss
0    0.4900 0.7063        0.5500    0.7063
1    0.5000 0.6978        0.4000    0.7035
2    0.5000 0.6935        0.5000    0.7027
3    0.5700 0.6908        0.5500    0.7029
4    0.5700 0.6890        0.5500    0.7030
5    0.5700 0.6875        0.5500    0.7033
6    0.5700 0.6862        0.5500    0.7033
7    0.5700 0.6849        0.5500    0.7035
8    0.5700 0.6837        0.5500    0.7041
9    0.6000 0.6824        0.5500    0.7046
```

I checked in another custom code that the training metrics recorded by the History and CSVLogger callbacks actually reflect the final step metrics of each epoch.

Any guidance or fix would be appreciated!

**Comment by mehtamansi29** [comment]
Hi @arthurflor23 -

Thanks for reporting the issue. Use  `verbose=2` while training in `model.fit(...)` , it will resolve the error.
```
Epoch 1/10
10/10 - 3s - 272ms/step - accuracy: 0.5300 - loss: 0.7008 - val_accuracy: 0.5000 - val_loss: 0.6875
Epoch 2/10
10/10 - 1s - 80ms/step - accuracy: 0.5200 - loss: 0.6925 - val_accuracy: 0.5500 - val_loss: 0.6863
Epoch 3/10
10/10 - 0s - 22ms/step - accuracy: 0.5500 - loss: 0.6896 - val_accuracy: 0.5000 - val_loss: 0.6867
Epoch 4/10
10/10 - 0s - 41ms/step - accuracy: 0.5600 - loss: 0.6874 - val_accuracy: 0.5000 - val_loss: 0.6876
Epoch 5/10
10/10 - 0s - 25ms/step - accuracy: 0.5700 - loss: 0.6854 - val_accuracy: 0.5000 - val_loss: 0.6886
Epoch 6/10
10/10 - 0s - 39ms/step - accuracy: 0.5800 - loss: 0.6835 - val_accuracy: 0.5500 - val_loss: 0.6896
Epoch 7/10
10/10 - 1s - 51ms/step - accuracy: 0.5800 - loss: 0.6817 - val_accuracy: 0.5500 - val_loss: 0.6906
Epoch 8/10
10/10 - 0s - 26ms/step - accuracy: 0.6000 - loss: 0.6800 - val_accuracy: 0.5500 - val_loss: 0.6915
Epoch 9/10
10/10 - 0s - 20ms/step - accuracy: 0.6000 - loss: 0.6784 - val_accuracy: 0.5000 - val_loss: 0.6923
Epoch 10/10
10/10 - 0s - 31ms/step - accuracy: 0.6000 - loss: 0.6768 - val_accuracy: 0.5000 - val_loss: 0.6930

Training History:
   accuracy      loss  val_accuracy  val_loss
0      0.53  0.700822          0.50  0.687477
1      0.52  0.692549          0.55  0.686306
2      0.55  0.689633          0.50  0.686686
3      0.56  0.687422          0.50  0.687614
4      0.57  0.685395          0.50  0.688602
5      0.58  0.683472          0.55  0.689624
6      0.58  0.681702          0.55  0.690604
7      0.60  0.680027          0.55  0.691548
8      0.60  0.678385          0.50  0.692278
9      0.60  0.676781          0.50  0.693044
```

Attached [gist](https://colab.sandbox.google.com/gist/mehtamansi29/9294e8efdc26e78dd48c5ad0657128ee/20212-mismatch-between-training-progress-and-history-csvlogger-callback-values.ipynb) for the reference.

[Verbosity](https://keras.io/api/models/model_training_apis/#fit-method) means control the level of information during training. This is for faster progressbar updates. 
[history](https://keras.io/guides/training_with_built_in_methods/#api-overview-a-first-endtoend-example) object is dictionary which store all metrics(accuracy,loss,validation accuracy, validation loss). These values are more accurate values.

**Comment by arthurflor23** [comment]
> Use `verbose=2` while training in `model.fit(...)` , it will resolve the error.

It worked; however, is the mismatch with `verbose=1` an issue, or is it expected?

I’m asking because, in our R&D project, we have training sessions where a single epoch takes hours to complete. So, using `verbose=2` feels like training in the dark for hours, with no feedback until we check the metrics at the end of the epoch to make a decision. On the other hand, `verbose=1` offers real-time monitoring, but without a detailed log afterward.

**Comment by mehtamansi29** [comment]
Hi @arthurflor23 -

`verbose=1` will display progress bar with training and validation. While `verbose=2` will shows summary metrics of training and validation after the epoch. Here you can find more details regarding [verbose](https://saturncloud.io/blog/understanding-the-use-of-verbose-in-keras-model-validation/#:~:text=verbose%3D1%20%3A%20Progress%20bar%20mode,validation%20metrics%20after%20each%20epoch.) in keras.

**Comment by arthurflor23** [comment]
> `verbose=1` will display progress bar with training and validation. While `verbose=2` will shows summary metrics of training and validation after the epoch. Here you can find more details regarding [verbose](https://saturncloud.io/blog/understanding-the-use-of-verbose-in-keras-model-validation/#:~:text=verbose%3D1%20%3A%20Progress%20bar%20mode,validation%20metrics%20after%20each%20epoch.) in keras.

I see, so it's expected. Thanks!

**Comment by arthurflor23** [comment]
As an additional comment, the `ModelCheckpoint` callback behaves incorrectly if there is no validation during training. In such cases, the callback uses the metric from the last step of an epoch to decide whether to save the model, even when the `frequency` parameter is set to `epoch`. So, in some cases, even if the average metric for the epoch improves, the model may not be saved because of the metric at the last step.
